File: feature-venues-history/src/main/java/home/oleg/placesnearme/feature_venues_history/di/VenuesHistoryModule.java
package home.oleg.placesnearme.feature_venues_history.di;

import android.arch.lifecycle.ViewModelProvider;
import android.arch.lifecycle.ViewModelProviders;
import android.support.v7.util.DiffUtil;

import com.smedialink.feature_add_favorite.CreateFavoriteViewModel;

import dagger.Module;
import dagger.Provides;
import home.oleg.placesnearme.core_presentation.recyclerview.ItemViewType;
import home.oleg.placesnearme.core_presentation.recyclerview.VenuesDiffCallback;
import home.oleg.placesnearme.feature_venues_history.HistoryVenuesAdapter;
import home.oleg.placesnearme.feature_venues_history.VenuesHistoryViewModel;
import home.oleg.placesnearme.feature_venues_history.VenuesHistoryFragment;
import io.reactivex.annotations.NonNull;

@Module
public abstract class VenuesHistoryModule {

    @Provides
    @NonNull
    static HistoryVenuesAdapter provideFavoriteVenuesAdapter(
            VenuesHistoryFragment fragment) {
        return new HistoryVenuesAdapter(VenuesDiffCallback.VENUES_DIFF_CALLBACK, fragment);
    }

    @Provides
    @NonNull
    static VenuesHistoryViewModel provideFavoritePlacesViewModel(
            VenuesHistoryFragment fragment,
            ViewModelProvider.Factory factory) {
        return ViewModelProviders.of(fragment, factory).get(VenuesHistoryViewModel.class);
    }

    @Provides
    @NonNull
    static CreateFavoriteViewModel provideCreateFavoriteViewModel(
            VenuesHistoryFragment fragment,
            ViewModelProvider.Factory factory) {
        return ViewModelProviders.of(fragment, factory).get(CreateFavoriteViewModel.class);
    }

}

File: feature-venues-history/src/main/java/home/oleg/placesnearme/feature_venues_history/di/VenuesHistoryModule.kt
package home.oleg.placesnearme.feature_venues_history.di

import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.ViewModelProviders
import com.smedialink.common.recyclerview.ItemsDiffCallback
import dagger.Module
import dagger.Provides
import dagger.multibindings.IntoMap
import home.oleg.placesnearme.coredi.mapkeys.ViewModelKey
import home.oleg.placesnearme.feature_add_favorite.presentation.CreateFavoriteViewModelDelegate
import home.oleg.placesnearme.feature_venues_history.domain.interactor.ObserveHistory
import home.oleg.placesnearme.feature_venues_history.presentation.VenuesHistoryViewModel
import home.oleg.placesnearme.feature_venues_history.presentation.ui.HistoryVenuesAdapter
import home.oleg.placesnearme.feature_venues_history.presentation.ui.VenuesHistoryFragment

@Module
object VenuesHistoryModule {

    @JvmStatic
    @Provides
    internal fun provideFavoriteVenuesAdapter(
            fragment: VenuesHistoryFragment): HistoryVenuesAdapter {
        return HistoryVenuesAdapter(ItemsDiffCallback.ITEMS_DIFF_CALLBACK, fragment)
    }

    @JvmStatic
    @IntoMap
    @ViewModelKey(VenuesHistoryViewModel::class)
    @Provides
    internal fun provideViewModel(
            createFavoriteViewModelDelegate: CreateFavoriteViewModelDelegate,
            observeHistory: ObserveHistory): ViewModel {
        return VenuesHistoryViewModel(createFavoriteViewModelDelegate, observeHistory)
    }

    @JvmStatic
    @Provides
    internal fun provideFavoritePlacesViewModel(
            fragment: VenuesHistoryFragment,
            factory: ViewModelProvider.Factory): VenuesHistoryViewModel {
        return ViewModelProviders.of(fragment, factory).get(VenuesHistoryViewModel::class.java)
    }

}
